rfcs/0000-dsl-driven-rda-code-generation.md

```diff
@@ -0,0 +1,876 @@
+
+
+## Table of Contents
+[Table of Contents]: #table-of-contents
+
+* [RFC Proposal](#rfc-proposal)
+* [Table of Contents](#table-of-contents)
+* [Motivation](#motivation)
+* [Proposed Solution](#proposed-solution)
+    * [Proposed Solution: Detailed Design](#proposed-solution-detailed-design)
+    * [Proposed Solution: Unresolved Questions](#proposed-solution-unresolved-questions)
+    * [Proposed Solution: Drawbacks](#proposed-solution-drawbacks)
+    * [Proposed Solution: Notable Alternatives](#proposed-solution-notable-alternatives)
+* [Prior Art](#prior-art)
+* [Future Possibilities](#future-possibilities)
+* [Addendums](#addendums)
+
+## Motivation
+[Motivation]: #motivation
+
+The RDA API is evolving rapidly and adding new fields as it moves towards production.
+Even after a production release is completed more fields will be added rapidly as that API development evolves from a first release focused on reliability and features to followup releases filling in more and more of the data available in the backend systems.
+
+Initially code to handle RDA API data was handwritten as there were relatively few fields at that time and the RDA API team had not yet established conventions for mapping the data into protobuf.
+Now the number of fields is growing and those conventions are well established.
+When estimating the ultimate size of the RDA API message objects the team indicated that there might ultimately be 2-5 times as many fields as now.
+
+With handwritten code the addition of new fields by the RDA API team can trigger many code changes across the project.
+These can include:
+- SQL migration scripts.
+- Hibernate database entity classes.
+- Data transformation/validation code to copy data from protobuf messages into entity objects.
+- Random synthetic data generation classes.
+- Data transformation code to copy data from Synthea data files into protobuf messages.
+
+Changes made to each of these areas require careful attention to ensure the logic, data types, and validation rules are correct.
+These changes have to be made consistently in different places in the code.
+And yet most of this code is repetitive since the fields follow established conventions.
+For example, every maximum field length in the database must be properly reflected in the database entities, enforced in the data validation code, and honored in the synthetic data generators.
+This can certainly be done with handwritten code but is error-prone and requires developer time to write/modify the code and review the associated PR.
+
+This RFC proposes replacing all of the handwritten code with a code generator that creates code using metadata in a YAML file.
+The code generator would remove the need to write and maintain large amounts of code.
+In the proof of concept the code generator replaced seven handwritten entity classes (over 1,000 LOC) and two transformer classes (over 1,400 lines) with a single 741 line YAML file.
```

Comment:
Also, I'm guessing this would replace several hundred lines of handwritten unit tests right?